__Name:__ Chowdhury Walid Tahabee

__UNI:__ ct3027

__Course:___ APAN5210: Python for Data Analysis

__Date:__ _03/13/2021_

## Mid Term Exam Extra Credit Assignment

### Code for Questions 1-7

In [1]:
class ColumbiaCourse():
    def __init__(self, CourseID, CourseName, Term):
        self.__CourseID=CourseID
        self.__CourseName=CourseName
        self.__Term=Term
    def getCourseID(self):
        return self.__CourseID
    def getCourseName(self):
        return self.__CourseName
    def setCourseStartDate(self,startdate):
        self.__CourseStartDate=startdate
    def getCourseStartDate(self):
        return self.__CourseStartDate
    
class SPSCourse(ColumbiaCourse):
    def __init__(self, CID, CName, CTerm, CourseLoc, CourseLec, CourseLecAssist):
        ColumbiaCourse(CID, CName, CTerm)
        self.__CourseLocation=CourseLoc
        self.__CourseLecturer=CourseLec
        self.__CourseLecturerAssistant=CourseLecAssist
    def setCourseLocation(self, CLoc):
        self.__CourseLocation=CLoc
    def getCourseLocation(self):
        return self.__CourseLocation
    def setCourseLecturer(self, CourseLec):
        self.__CourseLecturer=CourseLec
    def getCourseLecturer(self):
        return self.__CourseLecturer

### 1.	Which of the following is True about the __init__ method defined in Class ColumbiaCourse? 

A.	It displays the properties of the Class ColumbiaCourse

B.	It is a Value returning method 

___C.	It is a Constructor method for that class___

D.	None of the above

### 2.	Which of the following represents a template, blueprint, or contract that defines objects of      the same type?                             

___A. Class Definition___

B. A returning value method

C. Class instance

D. None of the above

### 3.	The constructor method of the child class SPSCourse invokes the constructor of parent class, ColumbiaCourse. 

A.	True

***B.	False***

### 4. We want to create an instance of  class ColumbiaCourse for our course. How would you go about doing this? Please select correct answer. 

In [2]:
#A.
Answer = 'A'
try:
    ColumbiaCourse('APANPS5210')
    ColumbiaCourse('Spring-2021')
    print(Answer+' is the correct answer!')
except Exception as ex:
    template = Answer+" results in {0}! Because, \n{1!r}"
    message = template.format(type(ex).__name__, ex.args)
    print(message)
    print(Answer+" is the WRONG ANSWER!")

A results in TypeError! Because, 
("__init__() missing 2 required positional arguments: 'CourseName' and 'Term'",)
A is the WRONG ANSWER!


In [3]:
#B.
Answer = 'B'
try:
    ColumbiaCourse('APANPS5210')
    ColumbiaCourse('Python for Data Analysis')
    print(Answer+' is the correct answer!')
except Exception as ex:
    template = Answer+" results in {0}! Because, \n{1!r}"
    message = template.format(type(ex).__name__, ex.args)
    print(message)
    print(Answer+" is the WRONG ANSWER!")

B results in TypeError! Because, 
("__init__() missing 2 required positional arguments: 'CourseName' and 'Term'",)
B is the WRONG ANSWER!


In [4]:
#C. 
Answer = 'C'
try:
    ColumbiaCourse('APANPS5210', 'Python For Data Analysis', 'Spring-2021')
    print(Answer+' is the correct answer!')
except Exception as ex:
    template = Answer+" results in {0}! Because, \n{1!r}"
    message = template.format(type(ex).__name__, ex.args)
    print(message)
    print(Answer+" is the WRONG ANSWER!")

C is the correct answer!


A.	ColumbiaCourse(‘APANPS5210’) and Then ColumbiaCouse(‘Spring-2021’)

B.	ColumbiaCourse(‘APANPS5210’) and Then ColumbiaCouse(‘Python For Data Analysis’)

___C.	ColumbiaCourse(‘APANPS5210’, ‘Python For Data Analysis’, ‘Spring-2021’)___

D.	None of the Above

### 5.	Suppose c1 is an instance of class ColumbiaCourse. Which code would you use to retrive the course name ? (1 pt)            

In [5]:
c1 = ColumbiaCourse('APAN5210', 'Python for Data Analysis', 'Spring 2021')

In [6]:
#A. 
Answer = 'A'
try:
    c1.getCourseName('APAN5210')
    print(Answer+' is the correct answer!')
except Exception as ex:
    template = Answer+" results in {0}! Because, \n{1!r}"
    message = template.format(type(ex).__name__, ex.args)
    print(message)
    print(Answer+" is the WRONG ANSWER!")

A results in TypeError! Because, 
('getCourseName() takes 1 positional argument but 2 were given',)
A is the WRONG ANSWER!


In [7]:
#B. 
Answer = 'B'
try:
    c1.getCourseName()
    print(Answer+' is the correct answer!')
except Exception as ex:
    template = Answer+" results in {0}! Because, \n{1!r}"
    message = template.format(type(ex).__name__, ex.args)
    print(message)
    print(Answer+" is the WRONG ANSWER!")

B is the correct answer!


A.  c1.getCourseName(‘APAN5210’)

___B.  c1.getCourseName()___

C.  Both A & B

D.  None of the above. 


### 6.	Now, you want to set the Course Start Date.  You will do this by using the following: c1.setCourseStartDate(‘1-11-2021’). Printing this command will give you the birthday.  (T or F?) (1 pt)    

In [8]:
c1.setCourseStartDate('1-11-2021')

In [9]:
print(c1.setCourseStartDate('1-11-2021'))

None


A.	True

___B.	False___

### 7.	You want to add a SPSCourse and the data for this instance include: Course ID is ‘APANPS5211’, Course Name is ‘Machine Learning’, Course Term is ‘Summer 2021’, Course Location is ‘LEW-Room 212’, Course Lecturer is ‘Michael Chen’, Course Lec Assistant is ‘Susan Ling’. How would you go about doing this?

In [10]:
#A.
Answer = "A"
try:
    ColumbiaCourse('APANPS5211', 'Machine Learning', 'Summer – 2020', 'LEW-Room 212', 'Michael Chen', 'Susan Ling')
    print(Answer+' is the correct answer!')
except Exception as ex:
    template = Answer+" results in {0}! Because, \n{1!r}"
    message = template.format(type(ex).__name__, ex.args)
    print(message)
    print(Answer+" is the WRONG ANSWER!")

A results in TypeError! Because, 
('__init__() takes 4 positional arguments but 7 were given',)
A is the WRONG ANSWER!


In [11]:
#B.
Answer = "B"
try:
    ColumbiaCourse('APANPS5211', 'Machine Learning', 'Summer – 2021')
    SPSCourse('LEW-Room 212', 'Michael Chen', 'Susan Ling')
    print(Answer+' is the correct answer!')
except Exception as ex:
    template = Answer+" results in {0}! Because, \n{1!r}"
    message = template.format(type(ex).__name__, ex.args)
    print(message)
    print(Answer+" is the WRONG ANSWER!")

B results in TypeError! Because, 
("__init__() missing 3 required positional arguments: 'CourseLoc', 'CourseLec', and 'CourseLecAssist'",)
B is the WRONG ANSWER!


In [12]:
#C.
Answer = "C"
try:
    SPSCourse('APANPS5211', 'Machine Learning', 'Summer – 2021', 'LEW-Room 212', 'Michael Chen', 'Susan Ling')
    print(Answer+' is the correct answer!')
except Exception as ex:
    template = Answer+" results in {0}! Because, \n{1!r}"
    message = template.format(type(ex).__name__, ex.args)
    print(message)
    print(Answer+" is the WRONG ANSWER!")

C is the correct answer!


A.	ColumbiaCourse('APANPS5211', 'Machine Learning’, ‘Summer – 2021’, ‘LEW-Room 212’, ‘Michael Chen’, ‘Susan Ling’)

B.	ColumbiaCourse(‘APANPS5211’, 'Machine Learning’, ‘Summer – 2021’) and then SPSCourse (‘LEW-Room 212’, ‘Michael Chen’, ‘Susan Ling’)

__C.	SPSCourse('APANPS5211', 'Machine Learning’, ‘Summer – 2021’, ‘LEW-Room 212’, ‘Michael Chen’, ‘Susan Ling’)__

D.	None of the above


### Code for Questions 8-10

In [13]:
class Product():
    def __init__(self,ProductName,ProductPrice):
        self.Name = ProductName.split("-")[0]
        self.Price= ProductPrice
        
    def Discount(self,percent):
        self.Price = self.Price-self.Price*percent
        return self.Price
        
    def Surcharge(self,amount):
        self.Price  += amount
        return self.Price
    
    def Fees(self):
        if self.Price < 5:
            fees = 0.1*self.Price
        else:
            fees = 0.15*self.Price
        self.Price += fees
        return self.Price


### 8.	Which is the output of the code below?

In [14]:
product1 = Product('PVC-Pipe',5)
product1.Discount(0.01)
product1.Fees()
print("Price for",product1.Name,"is $",product1.Price)

Price for PVC is $ 5.445


A.	Price for PVC-Pipe is $ 5

B.	Price for PVC Pipe is $ 5.5

___C.	Price for PVC is $ 5.445___

D.	Price for PVC is $ 5.5

E.	Price for P is $ 5.445

### 9.	What is the output of the code below? (1 pt)        

In [15]:
product2 = Product('B&D cordless drill',45)
product2.Discount(0.02)
product2.Surcharge(5)
product2.Fees()
print("Price for",product2.Name,"is $",product2.Price)

Price for B&D cordless drill is $ 56.465


A.	Price for B&D cordless drill is $ 50

B.	Price for B&D-cordless-drill is $ 55

C.	Price for B&D-cordless is $ 55

___D.	Price for B&D cordless drill is $ 56.465___

E.	Price for B is $ 56.465


### 10.	If we want to add a fixed fee of `$2 only to the products with a price higher than `$20, which block of code should be used to replace the method "Fees"  in this class?

In [16]:
price = 21

Use the above code chunk to test for different prices

In [17]:
#A. 
Answer = "A"
class Product():
    def __init__(self,ProductName,ProductPrice):
        self.Name = ProductName.split("-")[0]
        self.Price= ProductPrice
        
    def Discount(self,percent):
        self.Price = self.Price-self.Price*percent
        return self.Price
        
    def Surcharge(self,amount):
        self.Price  += amount
        return self.Price
    
    def Fees(self):
        if Price > 20:
            fees = 2
        else:
            fees=0
        Price = Price+fees
        return self.Price
    
try:
    test_product = Product('test_product',price)
    test_product.Fees()
    
    fee = int(test_product.Price - price)
    
    print("If price = $",price,"fees = $", fee)
    
    print(Answer+" is the correct answer!")
    
except Exception as ex:
    template = Answer+" results in {0}! Because, \n{1!r}"
    message = template.format(type(ex).__name__, ex.args)
    print(message)
    print(Answer+" is the WRONG ANSWER!")

A results in UnboundLocalError! Because, 
("local variable 'Price' referenced before assignment",)
A is the WRONG ANSWER!


In [18]:
#B. 
Answer = "B"
class Product():
    def __init__(self,ProductName,ProductPrice):
        self.Name = ProductName.split("-")[0]
        self.Price= ProductPrice
        
    def Discount(self,percent):
        self.Price = self.Price-self.Price*percent
        return self.Price
        
    def Surcharge(self,amount):
        self.Price  += amount
        return self.Price
    
    def Fees(self):
        if Price > 20:
            fees = 2
        else:
            fees=0
        self.Price = self.Price+fees
        return self.Price

try:
    test_product = Product('test_product',price)
    test_product.Fees()
    
    fee = int(test_product.Price - price)
    
    print("If price = $",price,"fees = $", fee)
    
    print(Answer+" is the correct answer!")
    
except Exception as ex:
    template = Answer+" results in {0}! Because, \n{1!r}"
    message = template.format(type(ex).__name__, ex.args)
    print(message)
    print(Answer+" is the WRONG ANSWER!")

B results in NameError! Because, 
("name 'Price' is not defined",)
B is the WRONG ANSWER!


In [19]:
#C. 
Answer = "C"
class Product():
    def __init__(self,ProductName,ProductPrice):
        self.Name = ProductName.split("-")[0]
        self.Price= ProductPrice
        
    def Discount(self,percent):
        self.Price = self.Price-self.Price*percent
        return self.Price
        
    def Surcharge(self,amount):
        self.Price  += amount
        return self.Price
    
    def Fees(self):
        if self.Price > 20:
            fees = 2
        else:
            fees=0
        self.Price = self.Price+fees
        return Price
    
try:
    test_product = Product('test_product',price)
    test_product.Fees()
    
    fee = int(test_product.Price - price)
    
    print("If price = $",price,"fees = $", fee)
    
    print(Answer+" is the correct answer!")
    
except Exception as ex:
    template = Answer+" results in {0}! Because, \n{1!r}"
    message = template.format(type(ex).__name__, ex.args)
    print(message)
    print(Answer+" is the WRONG ANSWER!")

C results in NameError! Because, 
("name 'Price' is not defined",)
C is the WRONG ANSWER!


In [20]:
#D.
Answer = "D"
class Product():
    def __init__(self,ProductName,ProductPrice):
        self.Name = ProductName.split("-")[0]
        self.Price= ProductPrice
        
    def Discount(self,percent):
        self.Price = self.Price-self.Price*percent
        return self.Price
        
    def Surcharge(self,amount):
        self.Price  += amount
        return self.Price
    
    def Fees(self):
        if self.Price > 20:
            fees = 2
        else:
            fees=0
        self.Price = self.Price+fees
        return self.Price

try:
    test_product = Product('test_product',price)
    test_product.Fees()
    
    fee = int(test_product.Price - price)
    
    print("If price = $",price,"fees = $", fee)
    
    print(Answer+" is the correct answer!")
    
except Exception as ex:
    template = Answer+" results in {0}! Because, \n{1!r}"
    message = template.format(type(ex).__name__, ex.args)
    print(message)
    print(Answer+" is the WRONG ANSWER!")

If price = $ 21 fees = $ 2
D is the correct answer!


A.	def Fees(self):

    if Price > 20:
        fees = 2
    else:
        fees=0
    Price = Price+fees
    return self.Price

B.	def Fees(self):

    if Price > 20:
        fees = 2
    else:
        fees=0
    self.Price = self.Price+fees
    return self.Price

C.	def Fees(self):

    if self.Price > 20:
        fees = 2
    else:
        fees=0
    self.Price = self.Price+fees
    return Price


***D.	def Fees(self):***

    if self.Price > 20:
        fees = 2
    else:
        fees=0
    self.Price = self.Price+fees
    return self.Price
